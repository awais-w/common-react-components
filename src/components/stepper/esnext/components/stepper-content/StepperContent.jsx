import React from 'react';
import PropTypes from 'prop-types';

const getStageType = (index, activeStepIndex) => {
  if (index < activeStepIndex) {
    return 'previous';
  }
  if (index === activeStepIndex) {
    return 'current';
  }
  return 'later';
};

const StepperContent = ({
  children,
  activeStepIndex,
  shouldNavBarUpdateOnScroll,
  latestStepReachedIndex,
  onClickPreviousStep,
  onWaypointEnter,
  renderUpToLatestStageReached,
}) =>
  React.Children.map(children, (child, index) => {
    if (index <= (renderUpToLatestStageReached ? latestStepReachedIndex : activeStepIndex)) {
      return React.cloneElement(child, {
        key: `step-${index}`,
        stageType: getStageType(index, activeStepIndex),
        shouldNavBarUpdateOnScroll,
        stepIndex: index,
        activeStepIndex,
        onClickPreviousStep,
        onWaypointEnter,
      });
    }
    return null;
  });

StepperContent.propTypes = {
  /**
   * Set the current step ( Zero based index )
   */
  activeStepIndex: PropTypes.number.isRequired,
  /**
   * The latest stage we've reached up to
   */
  latestStepReachedIndex: PropTypes.number.isRequired,
  /**
   * One or more `<StepperStage/>` components.
   */
  children: PropTypes.node.isRequired,
  /**
   * Determines whether we render a waypoint for each stage so we can update NavBar on scrolling over them
   */
  shouldNavBarUpdateOnScroll: PropTypes.bool.isRequired,
  /**
   * Used to update our parent with a new activeStepID when we click on previous stages
   */
  onClickPreviousStep: PropTypes.func.isRequired,
  /**
   * When we scroll this stage into view, this parent function will be called to update NavBar
   */
  onWaypointEnter: PropTypes.func.isRequired,
  /**
   * Determines whether we should render just to activeIndex or latestStepReachedIndex
   */
  renderUpToLatestStageReached: PropTypes.bool.isRequired,
};

export default StepperContent;
