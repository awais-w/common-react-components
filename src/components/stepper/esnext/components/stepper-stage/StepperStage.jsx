/**
 * Our <Stepper/> component handles drawing content / headers
 * This just provides a cleaner interface to construct <Stepper/>
 */
import React, { PureComponent, Fragment } from 'react';
import PropTypes from 'prop-types';
import Waypoint from 'react-waypoint';
import { Element } from 'react-scroll';
import uuidv4 from 'uuid/v4';
import bem from '@argos/utils/esnext/bem';
import classNames from 'classnames';
import scrollTo from '../../helpers/StepperFunctions';
import './StepperStage.scss';

const cn = bem({ prefix: 'cmc-', block: 'stepper-stage' });

class StepperStage extends PureComponent {
  uniqueId;

  constructor(props) {
    super(props);
    this.uniqueId = uuidv4();
  }

  scrollToMe() {
    scrollTo({
      componentName: this.uniqueId,
      duration: this.props.allowNavigation ? 500 : 0,
      parentID: this.props.parentScrollContainer,
    });
  }

  componentDidMount() {
    this.scrollToMe();
  }

  componentDidUpdate(prevProps) {
    if (
      prevProps.activeStepIndex !== this.props.activeStepIndex &&
      this.props.activeStepIndex === this.props.stepIndex
    ) {
      this.scrollToMe();
    }
  }

  render() {
    const {
      // Provided as part of <StepperStage/> props
      title,
      subTitle,
      content,
      // Added by <Stepper />
      stageType,
      stepIndex,
      onWaypointEnter,
      shouldNavBarUpdateOnScroll,
      shouldShowStageTitle,
      allowNavigation,
      onClickPreviousStep,
    } = this.props;

    const classes = classNames(cn(), {
      disabled: stageType !== 'current',
      locked: !allowNavigation,
    });

    return (
      <Fragment>
        <div
          key={`step-${stepIndex}`}
          onClick={
            allowNavigation && stageType === 'previous' ? () => onClickPreviousStep({ desiredIndex: stepIndex }) : null
          }
          className={classes}
          name={this.uniqueId}>
          <Element key={`step-${stepIndex}`} name={`step-${stepIndex}`}>
            {/* TODO - Provide prop */}
            {shouldShowStageTitle && (
              <div className={cn('headings')}>
                <h1>{title}</h1>
                {subTitle && <h4> - {subTitle}</h4>}
              </div>
            )}
            <div className={cn('content')}>{content}</div>
          </Element>
        </div>

        {shouldNavBarUpdateOnScroll && (
          <Waypoint
            onEnter={() => {
              onWaypointEnter({ stepIndex });
            }}
          />
        )}
      </Fragment>
    );
  }
}

StepperStage.propTypes = {
  /**
   * Primary title for this step
   */
  title: PropTypes.string.isRequired,
  /**
   * Show whats been completed in a previous step
   */
  subTitle: PropTypes.string,
  /**
   * Content for the current stage being rendered
   */
  content: PropTypes.node.isRequired,
  /**
   * Set the current step ( Zero based index )
   */
  activeStepIndex: PropTypes.number,
  /**
   * Where abouts this step is located in relation to current ( So we can for example )
   * Provide the ability to navigate back to stages marked as "Previous"
   */
  stageType: PropTypes.oneOf(['current', 'previous', 'later']),
  /**
   * The index of this stage in relation to other <StepperStage />'s
   */
  stepIndex: PropTypes.number,
  /**
   * When we scroll this stage into view, this parent function will be called to update NavBar
   */
  onWaypointEnter: PropTypes.func,
  /**
   * Used to update our parent with a new activeStepID when we click on previous stages
   */
  onClickPreviousStep: PropTypes.func,
  /**
   * Helper function in <Stepper/> to scroll to a component given it's container ID
   */
  shouldNavBarUpdateOnScroll: PropTypes.bool,
  /**
   * Should we show a title above this current stage content
   */
  shouldShowStageTitle: PropTypes.bool,
  /**
   * The parent container within which the scrolling needs to happen.
   */
  parentScrollContainer: PropTypes.string,
  /**
   * Flag to indicate whether navigation is allowed from stepper stage or not.
   */
  allowNavigation: PropTypes.bool,
};

StepperStage.defaultProps = {
  subTitle: '',
  // provide default values for those provided by <Stepper/>
  activeStepIndex: 0,
  stepIndex: 0,
  stageType: 'current',
  onWaypointEnter: () => {},
  onClickPreviousStep: () => {},
  scrollTo: () => {},
  shouldNavBarUpdateOnScroll: true,
  shouldShowStageTitle: false,
  parentScrollContainer: '',
};

export default StepperStage;
