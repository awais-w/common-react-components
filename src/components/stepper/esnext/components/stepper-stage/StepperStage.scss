@import '@cmc/common-styles/src/scss/_variables.scss';

.cmc-stepper-stage {
  @include font-standard;
  min-height: 100%;
  padding: $double;
  &__headings {
    display: inline-block;

    h1,
    h4 {
      display: inherit;
    }
  }
  &__page-divider {
    margin-bottom: $large;
  }
}
