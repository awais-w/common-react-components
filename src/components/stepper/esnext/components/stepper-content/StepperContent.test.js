import React from 'react';
import { shallow } from 'enzyme';
import StepperContent from './StepperContent';
import StepperStage from '../stepper-stage';

const defaultProps = {
  activeStepIndex: 1,
};

const renderLatestProps = {
  activeStepIndex: 1,
  latestStepReachedIndex: 2,
  renderUpToLatestStageReached: true,
};

const setup = ({ props }) => {
  const component = shallow(
    <StepperContent {...props}>
      <StepperStage />
      <StepperStage />
      <StepperStage />
    </StepperContent>,
  );
  return {
    wrapper: component,
    getStage: (index) => component.find(StepperStage).at(index),
  };
};

describe('StepperContent', () => {
  describe('base test', () => {
    const component = setup({ props: defaultProps });
    it('should match default snapshot', () => {
      expect(component.wrapper).toMatchSnapshot();
    });

    it('exists', () => {
      expect(component.wrapper).toExist();
    });

    describe("render appropriate <StepperStage>'s", () => {
      it("should not render future <StepperStages>'s", () => {
        expect(component.getStage(2)).not.toExist();
      });
      it("should render previous and current <StepperStage>'s", () => {
        expect(component.getStage(0)).toExist();
        expect(component.getStage(1)).toExist();
      });
    });
  });

  describe('when renderUpToLatestStageReached is true', () => {
    const component = setup({ props: renderLatestProps });

    describe('it contains correct stageType in relation to current step', () => {
      it('stageType is later if after activeStepIndex', () => {
        expect(component.getStage(2).props().stageType).toEqual('later');
      });
      it('stageType is current if equal to activeStepIndex', () => {
        expect(component.getStage(1).props().stageType).toEqual('current');
      });
      it('stageType is previous if before activeStepIndex', () => {
        expect(component.getStage(0).props().stageType).toEqual('previous');
      });
    });
  });
});
