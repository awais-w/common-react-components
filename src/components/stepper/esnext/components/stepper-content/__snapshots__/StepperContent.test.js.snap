// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StepperContent base test should match default snapshot 1`] = `
Array [
  <StepperStage
    activeStepIndex={1}
    key="step-0/.0"
    onClickPreviousStep={[Function]}
    onWaypointEnter={[Function]}
    parentScrollContainer=""
    scrollTo={[Function]}
    shouldNavBarUpdateOnScroll={true}
    shouldShowStageTitle={false}
    stageType="previous"
    stepIndex={0}
    subTitle=""
  />,
  <StepperStage
    activeStepIndex={1}
    key="step-1/.1"
    onClickPreviousStep={[Function]}
    onWaypointEnter={[Function]}
    parentScrollContainer=""
    scrollTo={[Function]}
    shouldNavBarUpdateOnScroll={true}
    shouldShowStageTitle={false}
    stageType="current"
    stepIndex={1}
    subTitle=""
  />,
]
`;
